src/Catalog/Helpers/CatalogWriterHelper.cs

```diff
@@ -0,0 +1,164 @@
+        /// <param name="telemetryService">A telemetry service.</param>
+        /// <param name="logger">A logger.</param>
+        /// <returns>A task that represents the asynchronous operation.
+        /// The task result (<see cref="Task{TResult}.Result" />) returns the latest
+        /// <see cref="DateTime}" /> that was processed.</returns>
+        public static async Task<DateTime> WritePackageDetailsToCatalogAsync(
+            IPackageCatalogItemCreator packageCatalogItemCreator,
+            SortedList<DateTime, IList<FeedPackageDetails>> packages,
+            IStorage storage,
+            DateTime lastCreated,
+            DateTime lastEdited,
+            DateTime lastDeleted,
+            int maxDegreeOfParallelism,
+            bool? createdPackages,
+            bool updateCreatedFromEdited,
+            CancellationToken cancellationToken,
+            ITelemetryService telemetryService,
+            ILogger logger)
+        {
+            if (packageCatalogItemCreator == null)
+            {
+                throw new ArgumentNullException(nameof(packageCatalogItemCreator));
+            }
+
+            if (packages == null)
+            {
+                throw new ArgumentNullException(nameof(packages));
+            }
+
+            if (storage == null)
+            {
+                throw new ArgumentNullException(nameof(storage));
+            }
+
+            if (maxDegreeOfParallelism < 1)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(maxDegreeOfParallelism),
+                    string.Format(Strings.ArgumentOutOfRange, 1, int.MaxValue));
+            }
+
+            if (telemetryService == null)
+            {
+                throw new ArgumentNullException(nameof(telemetryService));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var writer = new AppendOnlyCatalogWriter(storage, telemetryService, Constants.MaxPageSize);
+
+            var lastDate = DetermineLastDate(lastCreated, lastEdited, createdPackages);
+
+            if (packages.Count == 0)
+            {
+                return lastDate;
+            }
+
+            // Flatten the sorted list.
+            var workItems = packages.SelectMany(
+                    pair => pair.Value.Select(
+                        details => new PackageWorkItem(pair.Key, details)))
+                .ToArray();
+
+            await workItems.ForEachAsync(maxDegreeOfParallelism, async workItem =>
+            {
+                workItem.PackageCatalogItem = await packageCatalogItemCreator.CreateAsync(
+                    workItem.FeedPackageDetails,
+                    workItem.Timestamp,
+                    cancellationToken);
+            });
+
+            lastDate = packages.Last().Key;
```

Comment:
Maybe this line moved closer to line 94? I do not think that the packages are changed line 95 - line 107 code block.

Reply:
This is existing code which just moved to another type.